cpp/src/groupby/sort/group_shift.cu

```diff
@@ -0,0 +1,158 @@
+ */
+template <bool ForwardShift, typename EdgeIterator>
+struct group_shift_fill_functor {
+  EdgeIterator group_edges_begin;
+  size_type offset;
+  size_type group_label, offset_to_edge;
+
+  group_shift_fill_functor(EdgeIterator group_edges_begin, size_type offset)
+    : group_edges_begin(group_edges_begin), offset(offset)
+  {
+  }
+
+  __device__ size_type operator()(size_type i)
+  {
+    if (ForwardShift) {  // offset > 0
+      group_label    = i / offset;
+      offset_to_edge = i % offset;
+    } else {  // offset < 0
+      group_label    = -i / offset;
+      offset_to_edge = i % offset + offset + 1;
+    }
+    return *(group_edges_begin + group_label) + offset_to_edge;
+  }
+};
+
+}  // namespace
+
+/**
+ * @brief Implementation of groupby shift
+ *
+ * Groupby shift is based on sort groupby. The first step is a global shift for `sorted_values`.
+ * The second step is to set the proper locations to `fill_values`.
+ *
+ * @tparam EdgeIterator Iterator type to the group edge list
+ *
+ * @param sorted_values values to be sorted, grouped by keys
+ * @param offset The off set by which to shift the input
+ * @param fill_value Fill value for indeterminable outputs
+ * @param group_bound_begin Beginning of iterator range of the list that contains indices to the
+ * group's boundary. For forward shifts, the indices point to the groups' left boundaries, and right
+ * boundaries otherwise
+ * @param num_groups The number of groups
+ * @param mr Device memory resource used to allocate the returned table's device memory
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @return Column where values are shifted each group
+ */
+template <bool ForwardShift, typename EdgeIterator>
+std::unique_ptr<column> group_shift_impl(column_view const& sorted_values,
+                                         size_type offset,
+                                         cudf::scalar const& fill_value,
+                                         EdgeIterator group_bound_begin,
+                                         std::size_t num_groups,
+                                         rmm::cuda_stream_view stream,
+                                         rmm::mr::device_memory_resource* mr)
+{
+  // Step 1: global shift
+  auto shift_func = [col_size = sorted_values.size(), offset] __device__(size_type idx) {
+    auto raw_shifted_idx = idx - offset;
+    return static_cast<uint32_t>(
+      raw_shifted_idx >= 0 and raw_shifted_idx < col_size ? raw_shifted_idx : SAFE_GATHER_IDX);
+  };
+  auto gather_iter_begin = cudf::detail::make_counting_transform_iterator(0, shift_func);
+
+  auto shifted = cudf::detail::gather(table_view({sorted_values}),
+                                      gather_iter_begin,
+                                      gather_iter_begin + sorted_values.size(),
+                                      out_of_bounds_policy::DONT_CHECK,
```

Comment:
As an optimization, NULLIFY could be used to fill nulls, if fill_value is a null.

Reply:
`NULLIFY` will only fill nulls at the column boundary. But here we need to fill nulls at boundaries of each group. Also using `DONT_CHECK` gives better performance to `NULLIFY`.

Reply:
To go beyond column boundary, `values.size()` could be passed instead of `0`.
For non-fixed types, it gathers the first value (memory allocated and used too). That may not be desired behavior.

Reply:
Would `copy_if_else()` on `values` with `offset` be a suitable algorithm for this?